financial-templates-lib/price-feed/BalancerPriceFeed.js

```diff
@@ -0,0 +1,62 @@
+const assert = require("assert");
+const { PriceFeedInterface } = require("./PriceFeedInterface");
+const { BlockHistory, PriceHistory } = require("./utils");
+
+// Gets balancer prices as well as historical prices
+class BalancerPriceFeed extends PriceFeedInterface {
+  constructor(logger, web3, getTime, abi, address, tokenIn, tokenOut, lookback) {
+    assert(tokenIn, "BalancerPriceFeed requires tokenIn");
+    assert(tokenOut, "BalancerPriceFeed requires tokenOut");
+    assert(lookback, "BalancerPriceFeed requires lookback");
+    super();
+    this.logger = logger;
+    this.web3 = web3;
+    this.getTime = getTime;
+
+    this.contract = new web3.eth.Contract(abi, address);
+    this.currentPrice = null;
+    this.lastUpdateTime = null;
+    this.tokenIn = tokenIn;
+    this.tokenOut = tokenOut;
+    this.lookback = lookback;
+    // Provide a getblock function which returns the latest value if no number provided.
+    this.blockHistory = BlockHistory(number => web3.eth.getBlock(number >= 0 ? number : "latest"));
+
+    // Add a callback to get price, error can be thrown from web3 disconection or maybe something else
+    // which affects the update call.
+    this.priceHistory = PriceHistory(async number => {
+      return this.contract.methods.getSpotPriceSansFee(this.tokenIn, this.tokenOut).call(number);
+    });
+  }
+  getHistoricalPrice(time) {
+    // We want the block and price equal to or before this time
+    const block = this.blockHistory.getClosestBefore(time);
+    if (!this.priceHistory.has(block.timestamp)) {
+      return null;
+    }
+    return this.priceHistory.get(block.timestamp);
+  }
+  getLastUpdateTime() {
+    return this.lastUpdateTime;
+  }
+  getCurrentPrice() {
+    // normally would bubble up errors, but this is not supposed to throw
+    try {
+      return this.priceHistory.currentPrice();
+    } catch (err) {
```

Comment:
Do you think it's possible for this code to throw some error that's not just a missing price?

I don't see as easy of a way to make this detect a missing price without catching all other errors, but that may be preferable if you can think of a way.

Reply:
i think i will need to change the underlying pricehistory class to get the behavior you want. basically return null if no price and throw for any other errors. currently i throw if there is no price, but i think thats incompatible with this higher level api.